akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -196,13 +196,18 @@ private[stream] object Timers {
             pull(in)
           }
         }
+
+        override def onUpstreamFinish(): Unit = {
+          if (!isAvailable(in)) completeStage()
+        }
       })
 
       setHandler(out, new OutHandler {
         override def onPull(): Unit = {
           if (isAvailable(in)) {
             push(out, grab(in))
-            pull(in)
+            if (isClosed(in)) completeStage()
+            else pull(in)
           } else {
             if (nextDeadline.isOverdue()) {
```

Comment:
I am not sure. This is two main states (original input available or not) with two sub-cases. Of course this can be organized in various permutatons, so this is more about signaling intent. Why do you think a 3-state approach where one state have two sub-cases is better?